spring-cloud-gcp-vision/src/main/java/org/springframework/cloud/gcp/vision/DocumentOcrTemplate.java

```diff
@@ -0,0 +1,250 @@
+ * stored on Google Cloud Storage.
+ *
+ * @author Daniel Zou
+ */
+public class DocumentOcrTemplate {
+
+	private static final Pattern OUTPUT_PAGE_PATTERN = Pattern.compile("output-\\d+-to-\\d+\\.json");
+
+	private final ImageAnnotatorClient imageAnnotatorClient;
+
+	private final Storage storage;
+
+	public DocumentOcrTemplate(
+			ImageAnnotatorClient imageAnnotatorClient,
+			Storage storage) {
+		this.imageAnnotatorClient = imageAnnotatorClient;
+		this.storage = storage;
+	}
+
+	/**
+	 * Runs OCR processing for a specified {@code document} and generates OCR output files
+	 * under the path specified by {@code outputFilePathPrefix}. One JSON output file is
+	 * produced for each page of the document.
+	 *
+	 * <p>
+	 * For example, if you specify an {@code outputFilePathPrefix} of
+	 * "gs://bucket_name/ocr_results/myDoc_", all the output files of OCR processing will be
+	 * saved under prefix, such as:
+	 *
+	 * <ul>
+	 * <li>gs://bucket_name/ocr_results/myDoc_output-1-to-1.json
+	 * <li>gs://bucket_name/ocr_results/myDoc_output-2-to-2.json
+	 * <li>gs://bucket_name/ocr_results/myDoc_output-3-to-3.json
+	 * </ul>
+	 *
+	 * <p>
+	 * Note: OCR processing operations may take several minutes to complete, so it may not be
+	 * advisable to block on the completion of the operation. One may use the returned
+	 * {@link ListenableFuture} to register callbacks or track the status of the operation.
+	 *
+	 * @param document The {@link GoogleStorageLocation} of the document to run OCR processing
+	 * @param outputFilePathPrefix The {@link GoogleStorageLocation} of a file, folder, or a
+	 *     bucket describing the path for which all output files shall be saved under
+	 *
+	 * @return A {@link ListenableFuture} allowing you to register callbacks or wait for the
+	 * completion of the operation.
+	 */
+	public ListenableFuture<DocumentOcrResultSet> runOcrForDocument(
+			GoogleStorageLocation document,
+			GoogleStorageLocation outputFilePathPrefix) {
+		if (!document.isFile()) {
+			throw new IllegalArgumentException(
+					"Provided document location is not a valid file location: " + document);
+		}
+
+		GcsSource gcsSource = GcsSource.newBuilder()
+				.setUri(document.uriString())
+				.build();
+
+		Blob documentBlob = this.storage.get(
+				BlobId.of(document.getBucketName(), document.getBlobName()));
+		if (documentBlob == null) {
+			throw new IllegalArgumentException(
+					"Provided document does not exist: " + document);
+		}
+
+		String contentType = documentBlob.getContentType();
+		InputConfig inputConfig = InputConfig.newBuilder()
+				.setMimeType(contentType)
+				.setGcsSource(gcsSource)
+				.build();
+
+		GcsDestination gcsDestination = GcsDestination.newBuilder()
+				.setUri(outputFilePathPrefix.uriString())
+				.build();
+
+		OutputConfig outputConfig = OutputConfig.newBuilder()
+				.setGcsDestination(gcsDestination)
+				.setBatchSize(1)
```

Comment:
Done. This field means it will produce 1 JSON file per page in the document. If unset, this will [default to 20 pages per JSON file](https://cloud.google.com/vision/docs/reference/rpc/google.cloud.vision.v1#google.cloud.vision.v1.OutputConfig). And the range of values is 1-100.

I set it to 1 (aka 1-page-per-json-file) because it seemed like an intuitive default for someone who doesn't want to think about it. However upon reinspection it might make sense to make it configurable by the user ? (I suppose someone with a 1000-page pdf wouldn't want to output 1000 json files.)

Reply:
Configurable would be nice as a follow-up non-urgent PR.

Reply:
I ultimately changed it to configurable b/c I realized this is an important feature, i.e. batching seems important because it will reduce the amount of I/O needed for large documents.